akka-persistence-query/src/main/scala/akka/persistence/query/typed/internal/EventsBySliceFirehose.scala

```diff
@@ -0,0 +1,687 @@
+      setHandler(firehoseInlet, firehoseHandler)
+      setHandler(catchupInlet, catchupHandler)
+
+      private def tryPushOutput(): Unit = {
+        def tryPushFirehoseValue(deduplicate: Boolean): Boolean =
+          firehoseHandler.value match {
+            case OptionVal.Some(env) =>
+              firehoseHandler.value = OptionVal.None
+              if (!(deduplicate && isDuplicate(env))) {
+                if (settings.verboseLogging && log.isDebugEnabled)
+                  log.debug(
+                    s"Firehose entityType [$entityType] sliceRange [$sliceRangeStr] consumer [$consumerId] push from " +
+                    s" firehose [${env.persistenceId}] seqNr [${env.sequenceNr}], source [${env.source}]")
+                push(out, env)
+                true
+              } else
+                false
+            case _ =>
+              false
+          }
+
+        def tryPushCatchupValue(deduplicate: Boolean): Boolean =
+          catchupHandler.value match {
+            case OptionVal.Some(env) =>
+              catchupHandler.value = OptionVal.None
+              if (!(deduplicate && isDuplicate(env))) {
+                if (settings.verboseLogging && log.isDebugEnabled)
+                  log.debug(
+                    s"Firehose entityType [$entityType] sliceRange [$sliceRangeStr] consumer [$consumerId] push from " +
+                    s"catchup [${env.persistenceId}] seqNr [${env.sequenceNr}], source [${env.source}]")
+                push(out, env)
+                true
+              } else
+                false
+            case _ =>
+              false
+          }
+
+        if (isAvailable(out)) {
+          mode match {
+            case FirehoseOnly =>
+              // there can be one final value from catchup when switching to FirehoseOnly
+              if (!tryPushCatchupValue(deduplicate = false))
+                tryPushFirehoseValue(deduplicate = false)
+            case Both(_) =>
+              if (!tryPushFirehoseValue(deduplicate = true))
+                tryPushCatchupValue(deduplicate = true)
+            case CatchUpOnly =>
+              tryPushCatchupValue(deduplicate = false)
+          }
+        }
+
+        if (willShutDown) completeStage()
+      }
+
+      def isDuplicate(env: EventEnvelope[Any]): Boolean = {
+        if (settings.deduplicationCapacity == 0)
+          false
+        else {
+          val entry = DeduplicationCacheEntry(env.persistenceId, env.sequenceNr, env.source)
+          val result = {
+            if (deduplicationCache.contains(entry)) {
+              true
+            } else {
+              deduplicationCache.add(entry)
+              false
+            }
+          }
+
+          if (deduplicationCache.size >= deduplicationCacheEvictThreshold) {
+            // weird that add modifies the instance but drop returns a new instance
+            deduplicationCache = deduplicationCache.drop(deduplicationCache.size - settings.deduplicationCapacity)
+          }
```

Comment:
This means that once we have filled the cache, we will drop for each new element, maybe that is fine, or should we do an amortized scheme of some kind and drop a bunch at once once hitting the limit and not have to do it per element?